### DISCUSSION TITLE
Does the pre-render callback still exist?

### DISCUSSION THREAD

**Discussion by bsneed** [sbooth/SFBAudioEngine#290] [Q&A] [answered]
Hi Stephen,

Hope you're doing well.  I saw in the commit history it's been almost 10 years since we last talked :D   I see the library is very different, in all positive ways.  I've been searching for a bit in case I missed it, but there used to be a `SetAudioPreRenderCallback` method on the player.  You could set it and it'd hit your callback with the left and right buffer arrays, along with the number of frames.

I'm guessing it's gone, but thought I'd see if there was something equivalent that maybe I was overlooking.  I had been using this way back when to feed samples into quartz composer visualizations.

Thanks!

**Comment by sbooth** [comment, marked as answer]
I'm glad you're trying out the "new" version!

The pre-render callback method is gone but you should be able to accomplish something similar by installing a tap on the audio engine's output node, or depending on your needs, installing a render observer on one of the `AUAudioUnit`s in the graph.

**Reply by bsneed** [reply]
Excellent, thanks Stephen!  I'll give that a shot.

**Reply by bsneed** [reply]
Also, would you be ok with me adding this capability back with a PR?

**Reply by sbooth** [reply]
Since installing taps is something built into `AVAudioEngine` I'm not sure that adding a pre-render hook to the library would add much functionality.
